From 98eca5257c1d3d62ebaaf0ed8ee262381f3dc256 Mon Sep 17 00:00:00 2001
From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Thu, 4 May 2017 14:50:31 +1000
Subject: [PATCH libinput] lid: re-init the event listener after removing it

We may call remove on it again and that segfaults if it hasn't been
initialized.

https://bugzilla.redhat.com/show_bug.cgi?id=1440927

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
---
 src/evdev-lid.c         | 23 ++++++++++++++++++++---
 src/evdev-mt-touchpad.c |  6 ++++++
 2 files changed, 26 insertions(+), 3 deletions(-)

diff --git a/src/evdev-lid.c b/src/evdev-lid.c
index 8db7f37c..b7ff065e 100644
--- a/src/evdev-lid.c
+++ b/src/evdev-lid.c
@@ -98,7 +98,9 @@ lid_switch_toggle_keyboard_listener(struct lid_switch_dispatch *dispatch,
 					   dispatch);
 	} else {
 		libinput_device_remove_event_listener(
-						      &dispatch->keyboard.listener);
+					&dispatch->keyboard.listener);
+		libinput_device_init_event_listener(
+					&dispatch->keyboard.listener);
 	}
 }
 
@@ -174,6 +176,17 @@ evdev_read_switch_reliability_prop(struct evdev_device *device)
 }
 
 static void
+lid_switch_remove(struct evdev_dispatch *evdev_dispatch)
+{
+	struct lid_switch_dispatch *dispatch = lid_dispatch(evdev_dispatch);
+
+	if (!dispatch->keyboard.keyboard)
+		return;
+
+	libinput_device_remove_event_listener(&dispatch->keyboard.listener);
+}
+
+static void
 lid_switch_destroy(struct evdev_dispatch *evdev_dispatch)
 {
 	struct lid_switch_dispatch *dispatch = lid_dispatch(evdev_dispatch);
@@ -197,7 +210,9 @@ lid_switch_pair_keyboard(struct evdev_device *lid_switch,
 	if (dispatch->keyboard.keyboard) {
 		if (bus_kbd != BUS_I8042)
 			return;
+
 		libinput_device_remove_event_listener(&dispatch->keyboard.listener);
+		libinput_device_init_event_listener(&dispatch->keyboard.listener);
 	}
 
 	dispatch->keyboard.keyboard = keyboard;
@@ -225,7 +240,9 @@ lid_switch_interface_device_removed(struct evdev_device *device,
 
 	if (removed_device == dispatch->keyboard.keyboard) {
 		libinput_device_remove_event_listener(
-				      &dispatch->keyboard.listener);
+					&dispatch->keyboard.listener);
+		libinput_device_init_event_listener(
+					&dispatch->keyboard.listener);
 		dispatch->keyboard.keyboard = NULL;
 	}
 }
@@ -271,7 +288,7 @@ lid_switch_sync_initial_state(struct evdev_device *device,
 struct evdev_dispatch_interface lid_switch_interface = {
 	lid_switch_process,
 	NULL, /* suspend */
-	NULL, /* remove */
+	lid_switch_remove,
 	lid_switch_destroy,
 	lid_switch_interface_device_added,
 	lid_switch_interface_device_removed,
diff --git a/src/evdev-mt-touchpad.c b/src/evdev-mt-touchpad.c
index a35688b1..73f0fdd8 100644
--- a/src/evdev-mt-touchpad.c
+++ b/src/evdev-mt-touchpad.c
@@ -1673,6 +1673,12 @@ tp_interface_device_removed(struct evdev_device *device,
 		tp->dwt.keyboard = NULL;
 	}
 
+	if (removed_device == tp->lid_switch.lid_switch) {
+		libinput_device_remove_event_listener(
+					&tp->lid_switch.lid_switch_listener);
+		tp->lid_switch.lid_switch = NULL;
+	}
+
 	if (tp->sendevents.current_mode !=
 	    LIBINPUT_CONFIG_SEND_EVENTS_DISABLED_ON_EXTERNAL_MOUSE)
 		return;
-- 
2.12.2

